### ISSUE TITLE
Incompatibilities with the OpenTracing interface

### ISSUE THREAD

**Issue by eric-christian** [elastic/apm-agent-ruby#754]
Hello,

We are currently in the process of moving from jaeger to APM and encountered some problems with our implementation of the OpenTracing interface and the APM opentracing-bridge.

Before i continue with the points we found, i would like to ask a question: What is the current state of the OpenTracing bridge feature? Is it still maintained? Is it worse to continue investing time? :)

So far we found the following points:
1. Added FORMAT_TEXT_MAP for injection/extraction
2. `start_active_span { /code block/ }`, when a code block is given, returns the result of the code block, not the scope
3. The methods `start_span` and `start_active_span` have to take the parameter `tags`, not `labels`
4. `span.set_tag` must be implemented and has to return self
5. `span.finish` - the parameter `end_time` is not deprecated. Also we had issues with the default `clock_end: Util.monotonic_micros`. When a span is started, the default start_time is set as `start_time = Util.micros(Time.now)`. Combining `Util.micros(Time.now)` with `Util.monotonic_micros` resulted in durations that looked like this big numbers `1586156283628015`. We didn't go into more details here and simply stopped mixing them together.
6. The parameter `child_of` is supposed to be the parent span or parent span_context of the newly created span. The extracted span_context from ElasticAPM is already changed to represent a child of the once injected context.
7. `tracer.extract` has to return a `OpenTracing::SpanContext`, not an `ElasticAPM::TraceContext`.

We were able to fix all of these points, at least to our understanding.
Here is our fork https://github.com/eric-christian/apm-agent-ruby/commits/master
Is a PR welcome?

What we didn't check is the handling of any of the OpenTracing and ElasticAPM specific tags or log keys. We'll do this at a later point in time.

## Expected behavior
We expected to be able to simply switch from one tracer implementation to another. That is the reason why we started to implement the interface of OpenTracing and not the interface of a specific tracer.

## Environment

- **OS**: OS X
- **Ruby version**: 2.6.3
- **APM Server version**: 7.6.2
- **Agent version**: 3.6.0
- **OpenTracing version**: 0.5.0

Best regards,
Eric

**Comment by estolfo** [comment]
Hi @eric-christian! Thanks for opening this issue and for the detailed description. We'll investigate your points, but in the meantime, we certainly do welcome PRs! Feel free to open one with the work you've done.
Thanks and we'll get back to you soon.

**Comment by estolfo** [comment]
Hi @eric-christian I've opened a PR that includes your fixes, with a few tweaks. Thanks so much for opening this issue and for sharing your solutions. If you would like to review, the PR, it can be found [here](https://github.com/elastic/apm-agent-ruby/pull/791).

**Comment by eric-christian** [comment]
Hi @estolfo, that is awesome! I'll have a look on monday. :)

We also found some more bugs. Here is a quick list, i'll push some updates on monday.

One bug was regarding the formation of tag values. We added a `value.inspect` if the value is of a type that is not on the whitelist.

Another one was introduced by us when for some reason `start_span` returns a `OpenTracing::Span::NOOP_INSTANCE` which failed to have a `to_headers` method during injection.

Then we noticed that `elastic_span` has a `@timestamp` which seems to be used to position the span in the UI but is set to `Util.monotonic_micros` when `elastic_span.start( clock_start )` is called. Which moves the span visually to the right, if we process an event that has already happened. `OpenTracing.start_active_span('something that has already happened', start_time).finish(end_time)`. We are now setting the timestamp as well, but we did not had the time to test this yet.

Finally, we noticed that if we have more than one sequential child within a span, the `@self_time` of the parent span only takes the duration of the last child into account. Which seems to be wrong, but we are not 100% sure about that. We changed the implementation of `child_duration` to take the duration of all sequential children into account. But the test were still green, so we added some more tests for this situation. However, we are not entirely sure if we are still running into the correct direction here. Or if this even matters. :)
Here are the tests we added:
```ruby
      def create_child
        Span.new(
          name: 'span',
          transaction: transaction,
          trace_context: nil,
          parent: subject
        )
      end

      it 'calculates self_time' do
        subject.start
        travel 100
        child = create_child.start
        travel 100
        child.stop
        travel 100
        subject.stop

        expect(child.self_time).to eq 100
        expect(subject.self_time).to eq 200
      end

      context 'with more then one sequential child' do # NEW
        it 'calculates self_time' do
          subject.start
          travel 100
          create_child.start.tap { travel 100 }.stop
          travel 100
          create_child.start.tap { travel 100 }.stop
          travel 100
          subject.stop

          expect(subject.self_time).to eq 300 # This one fails because self_time is 400 with the old implementation
        end
      end

      context 'with nested children' do # NEW
        it 'calculates self_time' do
          subject.start
          travel 100
          create_child.start.tap do
            travel 100
            create_child.start.tap { travel 100 }.stop
          end.stop
          travel 100
          subject.stop

          expect(subject.self_time).to eq 200
        end
      end
```
Ok, now i wrote more than i actually wanted. ;)
I'll see if i can find the time to open a new issue on monday. Or should i update this one?

Best regards,
Eric

**Comment by estolfo** [comment]
Hi @eric-christian Thanks again for the detailed response and for looking into the issues you've found. Would you mind opening a separate issue for the self_time question? That way, we can work on that in parallel and separately from the the OpenTracing adjustments. 

I've accepted your suggested change to `#inject` and I'll look into your second concern regarding the start/end times.
Please let us know about more issues as you find them. Thanks!

**Comment by eric-christian** [comment]
Hi @estolfo, thank you for your work! 👍 

Plz ignore the unstructured infos from my last post. It was late and i shouldn't have written them into this issue!
I'll open a new issue once we are done with testing our current fixes. :)

Best regards,
Eric

**Comment by mikker** [comment]
Great! Good news 😊
Thanks a bunch for helping out!
